libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -0,0 +1,553 @@
+	{"i64.shl", 0x86},
+	{"i64.shr_s", 0x87},
+	{"i64.shr_u", 0x88},
+	{"i64.rotl", 0x89},
+	{"i64.rotr", 0x8a},
+	{"i32.wrap_i64", 0xa7},
+	{"i64.extend_i32_s", 0xac},
+	{"i64.extend_i32_u", 0xad},
+};
+
+bytes opcode(Opcode _o)
+{
+	return bytes(1, uint8_t(_o));
+}
+
+bytes lebEncode(uint64_t _n)
+{
+	bytes encoded;
+	while (_n > 0x7f)
+	{
+		encoded.emplace_back(uint8_t(0x80 | (_n & 0x7f)));
+		_n >>= 7;
+	}
+	encoded.emplace_back(_n);
+	return encoded;
+}
+
+bytes lebEncodeSigned(int64_t _n)
+{
+	if (_n >= 0 && _n < 0x40)
+		return bytes(1, uint8_t(uint64_t(_n) & 0xff));
+	else if (-_n > 0 && -_n < 0x40)
+		return bytes(1, uint8_t(uint64_t(_n + 0x80) & 0xff));
+	else
+		return bytes(1, uint8_t(0x80 | uint8_t(_n & 0x7f))) + lebEncodeSigned(_n / 0x80);
+}
+
+}
+
+bytes BinaryTransform::run(Module const& _module)
+{
+	BinaryTransform bt;
+
+	for (size_t i = 0; i < _module.globals.size(); ++i)
+		bt.m_globals[_module.globals[i].variableName] = i;
+
+	size_t funID = 0;
+	for (FunctionImport const& fun: _module.imports)
+		bt.m_functions[fun.internalName] = funID++;
+	for (FunctionDefinition const& fun: _module.functions)
+		bt.m_functions[fun.name] = funID++;
+
+	bytes ret{0, 'a', 's', 'm'};
+	ret += bytes{1, 0, 0, 0};
+	ret += bt.typeSection(_module.imports, _module.functions);
+	ret += bt.importSection(_module.imports);
+	ret += bt.functionSection(_module.functions);
+	ret += bt.memorySection();
+	ret += bt.globalSection();
+	ret += bt.exportSection();
+	for (auto const& sub: _module.subModules)
+	{
+		// TODO should we prefix and / or shorten the name?
+		bytes data = BinaryTransform::run(sub.second);
+		size_t length = data.size();
+		ret += bt.customSection(sub.first, std::move(data));
+		bt.m_subModulePosAndSize[sub.first] = {ret.size() - length, length};
+	}
+	ret += bt.codeSection(_module.functions);
+	return ret;
+}
+
+bytes BinaryTransform::operator()(Literal const& _literal)
+{
+	return opcode(Opcode::I64Const) + lebEncodeSigned(_literal.value);
+}
+
+bytes BinaryTransform::operator()(StringLiteral const&)
+{
+	// TODO is this used?
+	yulAssert(false, "String literals not yet implemented");
+}
+
+bytes BinaryTransform::operator()(LocalVariable const& _variable)
+{
+	return opcode(Opcode::LocalGet) + lebEncode(m_locals.at(_variable.name));
+}
+
+bytes BinaryTransform::operator()(GlobalVariable const& _variable)
+{
+	return opcode(Opcode::GlobalGet) + lebEncode(m_globals.at(_variable.name));
+}
+
+bytes BinaryTransform::operator()(BuiltinCall const& _call)
+{
+	if (_call.functionName == "dataoffset")
+	{
+		string name = boost::get<StringLiteral>(_call.arguments.at(0)).value;
+		return opcode(Opcode::I64Const) + lebEncodeSigned(m_subModulePosAndSize.at(name).first);
+	}
+	else if (_call.functionName == "datasize")
+	{
+		string name = boost::get<StringLiteral>(_call.arguments.at(0)).value;
+		return opcode(Opcode::I64Const) + lebEncodeSigned(m_subModulePosAndSize.at(name).second);
+	}
+
+	bytes args = visit(_call.arguments);
```

Comment:
Is this needed for unreachable? If yes, why not do it before `dataoffest/datasize`?

Reply:
It is not needed for unreachable, but we have to avoid visiting the arguments for `dataoffset` / `datasize` because it is an opaque identifier we should not encode.